pkg/virtctl/imageupload/imageupload.go

```diff
@@ -0,0 +1,355 @@
+	noCreate               = false
+)
+
+// HTTPClientCreator is a function that creates http clients
+type HTTPClientCreator func(bool) *http.Client
+
+var httpClientCreatorFunc HTTPClientCreator
+
+// SetHTTPClientCreator allows overriding the default http client
+// useful for unit tests
+func SetHTTPClientCreator(f HTTPClientCreator) {
+	httpClientCreatorFunc = f
+}
+
+// SetDefaultHTTPClientCreator sets the http client creator back to default
+func SetDefaultHTTPClientCreator() {
+	httpClientCreatorFunc = getHTTPClient
+}
+
+func init() {
+	SetDefaultHTTPClientCreator()
+}
+
+// NewImageUploadCommand returns a cobra.Command for handling the the uploading of VM images
+func NewImageUploadCommand(clientConfig clientcmd.ClientConfig) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "image-upload",
+		Short:   "Upload a VM image to a PersistentVolumeClaim.",
+		Example: usage(),
+		Args:    cobra.ExactArgs(0),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			v := command{clientConfig: clientConfig}
+			return v.run(cmd, args)
+		},
+	}
+	cmd.Flags().BoolVar(&insecure, "insecure", insecure, "Allow insecure server connections when using HTTPS.")
+	cmd.Flags().StringVar(&uploadProxyURL, "uploadproxy-url", "", "The URL of the cdi-upload proxy service.")
+	cmd.MarkFlagRequired("uploadproxy-url")
+	cmd.Flags().StringVar(&pvcName, "pvc-name", "", "The destination PVC.")
+	cmd.MarkFlagRequired("pvc-name")
+	cmd.Flags().StringVar(&pvcSize, "pvc-size", "", "The size of the PVC to create (ex. 10Gi, 500Mi).")
+	cmd.Flags().StringVar(&storageClass, "storage-class", "", "The storage class for the PVC.")
+	cmd.Flags().StringVar(&accessMode, "access-mode", accessMode, "The access mode for the PVC.")
+	cmd.Flags().StringVar(&imagePath, "image-path", "", "Path to the local VM image.")
+	cmd.MarkFlagRequired("image-path")
+	cmd.Flags().BoolVar(&noCreate, "no-create", noCreate, "Don't attempt to create a new PVC.")
+	cmd.Flags().UintVar(&uploadPodWaitSecs, "wait-secs", uploadPodWaitSecs, "Seconds to wait for upload pod to start.")
+	cmd.SetUsageTemplate(templates.UsageTemplate())
+	return cmd
+}
+
+func usage() string {
+	usage := `  # Upload a local disk image to a newly created PersistentVolumeClaim:
+	virtctl image-upload --insecure --upload-proxy-url=https://cdi-uploadproxy.mycluster.com --pvc-name=upload-pvc --pvc-size=10Gi --image-path=/images/fedora28.qcow2`
+	return usage
+}
+
+type command struct {
+	clientConfig clientcmd.ClientConfig
+}
+
+func (c *command) run(cmd *cobra.Command, args []string) error {
+	file, err := os.Open(imagePath)
+	if err != nil {
+		return err
+	}
+	defer file.Close()
+
+	namespace, _, err := c.clientConfig.Namespace()
+	if err != nil {
+		return err
+	}
+
+	virtClient, err := kubecli.GetKubevirtClientFromClientConfig(c.clientConfig)
+	if err != nil {
+		return fmt.Errorf("Cannot obtain KubeVirt client: %v", err)
+	}
+
+	pvc, err := getUploadPVC(virtClient, namespace, pvcName, noCreate)
```

Comment:
he `noCreate` looks confusing to me. Here we pass in `noCreate = false/true`. and below we check if the pvc does not exist and create it, if not. See also the related comment at `getUploadPVC`.

Reply:
okay will refactor.  getUploadPVC will not modify PVC and will return error if not exists.  will add ensureUploadAnnotation() for case when PVC does exist